WebGL creating long dependencies in backboard render times

kkinnunen-apple opened the pull request:
#### db0a320e698a2411f40ab61f51e5dcd9f07bc8bf
<pre>
WebGL creating long dependencies in backboard render times
<a href="https://bugs.webkit.org/show_bug.cgi?id=254912">https://bugs.webkit.org/show_bug.cgi?id=254912</a>
rdar://106957008

Reviewed by Simon Fraser.

Before, WebGL would wait for 3rd oldest frame to finish before
continuing to create a new frame. In case the frames were very slow,
they would slow the whole system down as the system compositor
would wait for each submitted frame to finish.

Instead, wait at the WebKit level for the WebGL frames to finish
rendering before submitting the frames to the system compositor.

Wait WebGL rendering to IOSurfaces to finish before setting the surface
as the &quot;delegated contents&quot; of the PlatformCALayer.

Original plan was to wait for the MTLSharedEvent associated with the
context. This is not useful, as waiting has to happen in WP and using
MTLSharedEvent in WP is not possible in full GPUP mode. Instead,
encapsulate the wait fence functionality in new interface
PlatformCALayerDelegatedContentsFence.

The feature has client configurations:
 1. WP-side compositing, WP-side WebGL
 2. WP-side compositing, GPUP-side WebGL
 3. UI-side compositing, GPUP-side WebGL

WP-side compositing:
Background: delegated contents is being set during [CALayer display].

The fence is waited on during the [CALayer display]
when the delegated contents is being set to the PlatformCALayer. This
happens on the main thread.

UI-side compositing:
Background: delegated contents is being set during
RemoteLayerTreeContext::buildTransaction, prepareBackingStoresForDisplay
phase.

When setting the delegated contents, add the fence to the list
of &quot;front buffer flushers&quot;. The fence is wrapped inside a wrapper that
implements the flusher interface.

Waiting happens on the flusher dispatch queue.

WP-side WebGL
Background: During WebPage rendering update, the WebGL context layers
will be &quot;prepared for display&quot;.

As a last ANGLE OpenGL command for the frame, this will insert the EGL
fence that will signal the finishing of the frame. The inserted fence
(ANGLE EGL sync object) encapsulates a MTLSharedEvent signal. Upon
completion of the event signal, Metal will invoke the
MTLSharedEventListener with the pre-set block for the signal value, in
the Metal owned background thread. The block will notify the delegated
contents fence Condition.

GPUP-side WebGL
During prepare for display, send IPC::Semaphore to the GPUP side.
This semaphore will be waited on by the implementation of
PlatformCALayerDelegatedContentsFence.

In GPUP side, insert the semaphore signal into the EGL fence signal,
similar to above.
IPC::Semaphore is used to trampoline the notification, because
WP cannot use MTLSharedEvent directly due to IOKit blocking. IPC
message waits cannot be used because they would spin the message
processing, the IPC connection does not exist at the compositor level,
OffscreenCanvas would have different connection thread, and flusher
dispatch queue doesn&apos;t have the IPC connection.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::waitAndUpdateOldestFrame): Deleted.
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::setDelegatedContents):
(WebCore::PlatformCALayer::setDelegatedContentsFinishedEvent): Deleted.
* Source/WebCore/platform/graphics/ca/PlatformCALayerDelegatedContents.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::setDelegatedContents):
(WebCore::PlatformCALayerCocoa::setDelegatedContentsFinishedEvent): Deleted.
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newSharedEventWithMachPort):
(WebCore::newSharedEvent):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::MetalFenceSharedState::event):
(WebCore::GraphicsContextGLCocoa::MetalFenceSharedState::listener):
(WebCore::GraphicsContextGLCocoa::MetalFenceSharedState::MetalFenceSharedState):
(WebCore::GraphicsContextGLCocoa::platformInitialize):
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLCocoa::createSyncWithSharedEvent):
(WebCore::GraphicsContextGLCocoa::clientWaitSyncWithFlush):
(WebCore::GraphicsContextGLCocoa::prepareForDisplay):
(WebCore::GraphicsContextGLCocoa::prepareForDisplayWithFinishedSignal):
(WebCore::GraphicsContextGLCocoa::insertCommandsCompleteSignalOrInvoke):
(WebCore::GraphicsContextGLCocoa::metalFenceSharedState):
(WebCore::GraphicsContextGLCocoa::MetalFenceSharedState::create):
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
(): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
(WebKit::RemoteGraphicsContextGLCocoa::prepareForDisplay):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::setDelegatedContents):
(WebKit::RemoteLayerBackingStore::setDelegatedContentsFinishedEvent): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/RemoteGraphicsContextGLProxyCocoa.mm:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::setDelegatedContentsFinishedEvent): Deleted.

Canonical link: <a href="https://commits.webkit.org/263427@main">https://commits.webkit.org/263427@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/82b129bb32c2298e7fa93c49caf82ca974b2bd74

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4351 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4598 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5826 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4567 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4344 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4598 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4434 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4795 "1 flakes 143 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4412 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4570 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3916 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5825 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2066 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3893 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/7294 "220 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3896 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3960 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5499 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4367 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3549 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3888 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3891 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1127 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/7943 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4242 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/platform/graphics/ca/PlatformCALayerDelegatedContents.h:
```diff
@@ -29,33 +29,24 @@
+    RefPtr<PlatformCALayerDelegatedContentsFence> finishedFence;
 };
 
 struct PlatformCALayerInProcessDelegatedContents {
```
Is there a build configuration where both `PlatformCALayerDelegatedContents` and `PlatformCALayerInProcessDelegatedContents` are required in the same binary?

kkinnunen-apple replied:
Yeah, WebGL GPUP is toggleable on runtime, as is ui-side compositing

djg commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
```diff
@@ -26,6 +26,7 @@
 #pragma once
 
 #include "PlatformCALayer.h"
+#include "PlatformCALayerDelegatedContents.h"
```
Is this include required if the rest of the changes to header doesn't use it?

kkinnunen-apple replied:
You're right, it's not needed. Removed.

djg commented on Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
```diff
@@ -85,15 +85,28 @@ void destroyPbufferAndDetachIOSurface(EGLDisplay display, void* handle)
     EGLDeviceEXT device = EGL_NO_DEVICE_EXT;
     if (!EGL_QueryDisplayAttribEXT(display, EGL_DEVICE_EXT, reinterpret_cast<EGLAttrib*>(&device)))
-        return RetainPtr<MTLSharedEvent>();
+        return nullptr;
```
fair enough.

djg commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
```diff
@@ -122,6 +133,8 @@ class WEBCORE_EXPORT GraphicsContextGLCocoa : public GraphicsContextGLANGLE {
     IntSize m_mediaSampleRotationSessionSize;
 #endif
+    RetainPtr<MTLSharedEventListener> m_finishedMetalSharedEventListener;
+    RetainPtr<id> m_finishedMetalSharedEvent; // FIXME: Remove all C++ includees and use id<MTLSharedEvent>.
```
nit: includes

kkinnunen-apple replied:
This actually tries to use "includees" as "employees", e.g. "remove all files that include this file".

djg commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -76,6 +76,7 @@
     nullptr
 };
 
+
```
nit: remove extra blank line.

djg commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -964,6 +964,29 @@ static bool needsEAGLOnMac()
+void GraphicsContextGLCocoa::insertFinishedSignalOrInvoke(Function<void()> signal)
+{
+    if (!contextAttributes().useMetal) {
+        signal();
```
Is this the correct thing to do? It feels like `Function<void()> signal` is acting as a promise and the documentation says that signal is called with the rendering is finished, but in this case it isn't. Should the `signal` take a bool indicating whether rendering is finished, or not?

kkinnunen-apple replied:
the docs also say "If not possible, calls the `signal`."

E.g. there's nothing to be done if we don't have a primitive to do that thing being asked to do.

The intention is that the code is structured so that the caller obtains the previous behavior, e.g scheduled rendering, with the same code structure. This would be the logic the caller would anyway default to, if they knew "backend doesn't support singling on finished."

djg commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -60,6 +60,32 @@
 
 using namespace WebCore;
 
+constexpr Seconds delegatedContentsFinishedTimeout = 5_s;
```
This value is now cut-n-pasted into 3 locations. Does it require a note that if it ever changes to update the other values?

kkinnunen-apple replied:
The context timeout and the compositor timeout are a bit different. I'll unify the compositor ones and think about the context one.

djg commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -60,6 +60,32 @@
+public:
+    static std::unique_ptr<DelegatedContentsFenceFlusher> create(RefPtr<PlatformCALayerDelegatedContentsFence> fence)
+    {
+        return std::unique_ptr<DelegatedContentsFenceFlusher> { new DelegatedContentsFenceFlusher(WTFMove(fence)) };
```
nit: std::make_unique

kkinnunen-apple replied:
It doesn't work because the constructor is private, using it will cause more problems than what it solves (it doesn't really solve anything..)

djg commented on Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:
```diff
@@ -356,6 +384,35 @@ static bool hasMultipleGPUs()
+    context->markContextChanged();
+    context->clearColor(0.f, 1.f, 0.f, 1.f);
+    context->clear(WebCore::GraphicsContextGL::COLOR_BUFFER_BIT);
+    std::atomic<bool> gotSignal = false;
```
`signalled`? (`gotSignal` sounds horrible to my ear) `while not signalled` reads better than `while not got signal`.

djg commented on Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestGraphicsContextGLCocoa.mm:
```diff
@@ -356,6 +384,35 @@ static bool hasMultipleGPUs()
+    });
+    while (!gotSignal)
+        sleep(.1_s);
+    EXPECT_TRUE(gotSignal);
```
Is it possible for `prepareForDisplayWithFinishedSignal` to timeout? Does the possibility of timeout make this a flaky test?

kkinnunen-apple replied:
The function `prepareForDisplayWithFinishedSignal` itself is non-asynchronous, so it doesn't timeout.

The `signal` may be never invoked, which is a bug and which is then caught by the test. The test will hang, i.e. timeout then.

djg reviewed (approved):
:shipit:

kkinnunen-apple committed:
WebGL creating long dependencies in backboard render times
https://bugs.webkit.org/show_bug.cgi?id=254912
rdar://106957008

Reviewed by Simon Fraser.

Before, WebGL would wait for 3rd oldest frame to finish before
continuing to create a new frame. In case the frames were very slow,
they would slow the whole system down as the system compositor
would wait for each submitted frame to finish.

Instead, wait at the WebKit level for the WebGL frames to finish
rendering before submitting the frames to the system compositor.

Wait WebGL rendering to IOSurfaces to finish before setting the surface
as the "delegated contents" of the PlatformCALayer.

Original plan was to wait for the MTLSharedEvent associated with the
context. This is not useful, as waiting has to happen in WP and using
MTLSharedEvent in WP is not possible in full GPUP mode. Instead,
encapsulate the wait fence functionality in new interface
PlatformCALayerDelegatedContentsFence.

The feature has client configurations:
 1. WP-side compositing, WP-side WebGL
 2. WP-side compositing, GPUP-side WebGL
 3. UI-side compositing, GPUP-side WebGL

WP-side compositing:
Background: delegated contents is being set during [CALayer display].

The fence is waited on during the [CALayer display]
when the delegated contents is being set to the PlatformCALayer. This
happens on the main thread.

UI-side compositing:
Background: delegated contents is being set during
RemoteLayerTreeContext::buildTransaction, prepareBackingStoresForDisplay
phase.

When setting the delegated contents, add the fence to the list
of "front buffer flushers". The fence is wrapped inside a wrapper that
implements the flusher interface.

Waiting happens on the flusher dispatch queue.

WP-side WebGL
Background: During WebPage rendering update, the WebGL context layers
will be "prepared for display".

As a last ANGLE OpenGL command for the frame, this will insert the EGL
fence that will signal the finishing of the frame. The inserted fence
(ANGLE EGL sync object) encapsulates a MTLSharedEvent signal. Upon
completion of the event signal, Metal will invoke the
MTLSharedEventListener with the pre-set block for the signal value, in
the Metal owned background thread. The block will notify the delegated
contents fence Condition.

GPUP-side WebGL
During prepare for display, send IPC::Semaphore to the GPUP side.
This semaphore will be waited on by the implementation of
PlatformCALayerDelegatedContentsFence.

In GPUP side, insert the semaphore signal into the EGL fence signal,
similar to above.
IPC::Semaphore is used to trampoline the notification, because
WP cannot use MTLSharedEvent directly due to IOKit blocking. IPC
message waits cannot be used because they would spin the message
processing, the IPC connection does not exist at the compositor level,
OffscreenCanvas would have different connection thread, and flusher
dispatch queue doesn't have the IPC connection.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::waitAndUpdateOldestFrame): Deleted.
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::setDelegatedContents):
(WebCore::PlatformCALayer::setDelegatedContentsFinishedEvent): Deleted.
* Source/WebCore/platform/graphics/ca/PlatformCALayerDelegatedContents.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::setDelegatedContents):
(WebCore::PlatformCALayerCocoa::setDelegatedContentsFinishedEvent): Deleted.
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newSharedEventWithMachPort):
(WebCore::newSharedEvent):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::MetalFenceSharedState::event):
(WebCore::GraphicsContextGLCocoa::MetalFenceSharedState::listener):
(WebCore::GraphicsContextGLCocoa::MetalFenceSharedState::MetalFenceSharedState):
(WebCore::GraphicsContextGLCocoa::platformInitialize):
(WebCore::GraphicsContextGLANGLE::~GraphicsContextGLANGLE):
(WebCore::GraphicsContextGLCocoa::createSyncWithSharedEvent):
(WebCore::GraphicsContextGLCocoa::clientWaitSyncWithFlush):
(WebCore::GraphicsContextGLCocoa::prepareForDisplay):
(WebCore::GraphicsContextGLCocoa::prepareForDisplayWithFinishedSignal):
(WebCore::GraphicsContextGLCocoa::insertCommandsCompleteSignalOrInvoke):
(WebCore::GraphicsContextGLCocoa::metalFenceSharedState):
(WebCore::GraphicsContextGLCocoa::MetalFenceSharedState::create):
* Source/WebCore/platform/graphics/cocoa/WebProcessGraphicsContextGLCocoa.mm:
(): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLCocoa.cpp:
(WebKit::RemoteGraphicsContextGLCocoa::prepareForDisplay):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::setDelegatedContents):
(WebKit::RemoteLayerBackingStore::setDelegatedContentsFinishedEvent): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/RemoteGraphicsContextGLProxyCocoa.mm:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::setDelegatedContentsFinishedEvent): Deleted.

Canonical link: https://commits.webkit.org/263427@main
